craft_application/services/base.py

```diff
@@ -41,5 +46,32 @@ def __init__(self, app: AppMetadata, services: ServiceFactory) -> None:
         self._services = services
 
     def setup(self) -> None:
-        """Application-specific service setup."""
+        """Service-specific setup to perform.
+
+        Child classes should always call ``super().setup()``.
+        """
         emit.debug(f"Setting up {self.__class__.__name__}")
+
+    @property
+    def _project(self) -> Project:
+        """The rendered project to use.
+
+        This will error if the project is not yet fully configured.
+        """
+        return self._services.get("project").get()
+
+    @property
+    def _build_info(self) -> BuildInfo:
+        """The information about the current build.
+
+        This will error if the build plan is not configured, if the build plan contains
+        multiple entries, or if the build plan is empty.
+        """
+        plan = self._services.get("build_plan").plan()
+        match len(plan):
+            case 0:
+                raise errors.EmptyBuildPlanError
```

Comment:
`self._services.get("build_plan").plan()` is already raising an `EmptyBuildPlanError` if the plan is empty, so do we need to re-check it here?

Reply:
Strictly speaking no, but it makes the linters happy to include this case.